### ISSUE TITLE
cast/type issue in high_variance confounds

### ISSUE THREAD

**Issue by rschmaelzle** [nilearn/nilearn#896]
Hi All,
I am hitting a cast error with this line

```python
hv_confounds = nilearn.image.high_variance_confounds(func_filename)
```
The error goes like this
```
.....
 --> 144     signals -= np.mean(signals, axis=0)
    145     if type == "linear":
    146         # Keeping "signals" dtype avoids some type conversion further down,

TypeError: Cannot cast ufunc subtract output from dtype('float64') to dtype('<i2') with casting rule 'same_kind'
....
```

the versions we have installed are: nilearn 0.1.4.post1  and numpy 1.10.1
Any help is much appreciated! Love nilearn!
Best, Ralf.

PS see this seems to relate to https://github.com/scikit-learn/scikit-learn/issues/5397
PPS tried downgrading to previous numpy "conda install numpy=1.9.1", but then there are dependencies to scipy...

**Comment by rschmaelzle** [comment]
I see this probably relates to this: 
https://github.com/scikit-learn/scikit-learn/issues/5397
am trying to fix, but so far unsuccessful.

**Comment by lesteve** [comment]
> PS see this seems to relate to scikit-learn/scikit-learn#5397

What's your scikit-learn version? I would have thought this would be fixed in scikit-learn 0.17.

Conda tip: not sure exactly why, but you need to explicitly specify dependencies in conda when using conda install with a specific version, e.g. something like:
`conda install numpy=1.9.1 scipy scikit-learn'

otherwise you'll keep scipy and scikit-learn compiled with whatever your previous version of numpy was.

**Comment by rschmaelzle** [comment]
Thanks lesteve,
I tried to update to scikit learn 0.17, but this didn't solve the issue. 
Am still testing to get it to work via downgrading (thanks for the tip), which involved more dependency issues than scipy and scikit-learn.
Best, Ralf.

**Comment by lesteve** [comment]
Looking at this one more carefully, this doesn't have anything to do with scikit-learn. Here is a snippet to reproduce the issue:

```python
import numpy as np
import nibabel
from nilearn import image

img = nibabel.Nifti1Image(np.arange(54, dtype='i8').reshape(3, 3, 3, 2), np.eye(4))
image.high_variance_confounds(img)
```

I guess we need to convert int arrays into floats as we do in other places in nilearn.

**Comment by GaelVaroquaux** [comment]
> I guess we need to convert int arrays into floats as we do in other places in nilearn.

+1

**Comment by KamalakerDadi** [comment]
sklearn.utils.validation.as_float_array can be used here  to fix ?

**Comment by KamalakerDadi** [comment]
May be also related to #471

**Comment by GaelVaroquaux** [comment]
> sklearn.utils.validation.as_float_array can be used here to fix ?

Indeed.

**Comment by rschmaelzle** [comment]
Hi ,
in nilearn's signal.py function I changed the following:
      from sklearn.utils import as_float_array, check_array, check_symmetric
and 
in line 146 of the _detrend function, where the error was thrown...
       signals = as_float_array(signals)    
       signals -= np.mean(signals, axis=0)
That seems to fix it for me and for now. Happy to proceed with learning nilearn.
Best r

**Comment by GaelVaroquaux** [comment]
Fix submitted in #920 . The problem will be fixed in nilearn 0.2.2
